examples/utils/example-utils/src/migrator/sameContainerMigrator.ts

```diff
@@ -73,221 +81,175 @@ export class SameContainerMigrator
-			this.ensureMigrating,
-			this.ensureLoading,
-		);
+		const migratable = this._currentModel;
+		this._acceptedVersion = migratable.migrationTool.acceptedVersion;
+		if (this._acceptedVersion === undefined) {
+			throw new Error("Expect an accepted version before migration starts");
+		}
+		// It's possible that our modelLoader is older and doesn't understand the new acceptedVersion.
+		// Currently this fails the migration gracefully and emits an event so the app developer can know
+		// they're stuck. Ideally the app developer would find a way to acquire a new ModelLoader and move
+		// forward, or at least advise the end user to refresh the page or something.
+		// TODO: Does the app developer have everything they need to dispose gracefully when recovering with
+		// a new ModelLoader?
+		const migrationSupported = await this.modelLoader.supportsVersion(this._acceptedVersion);
+		if (!migrationSupported) {
+			this.emit("migrationNotSupported", this._acceptedVersion);
+			return;
+		}
+
+		this._detachedModel = await this.modelLoader.createDetached(this._acceptedVersion);
+		this._migratedModel = this._detachedModel.model;
 	};
 
-	private readonly ensureMigrating = () => {
-		// ensureMigrating() is called when we reach the "migrating" state. This should likely only happen once, but
-		// can happen multiple times if we disconnect during the migration process.
+	private readonly getExportedData = async () => {
+		const acceptedSeqNum = this.currentModel.migrationTool.acceptedSeqNum;
+		assert(acceptedSeqNum !== undefined, "acceptedSeqNum should be defined");
+		this._pausedModel = await this.modelLoader.loadExistingPaused(
+			this._currentModelId,
+			acceptedSeqNum,
+		);
+		assert(
+			this.pausedModel.container.deltaManager.lastSequenceNumber === acceptedSeqNum,
+			"paused model should be at accepted sequence number",
+		);
+		this._exportedData = await this.pausedModel.exportData();
+	};
 
-		if (!this.connected) {
-			// If we are not connected we should wait until we reconnect and try again. Note: we re-enter the state
-			// machine, since its possible another client has already completed the migration by the time we reconnect.
-			this.currentModel.once("connected", this.takeAppropriateActionForCurrentMigratable);
+	private readonly generateTransformedData = async () => {
+		// TODO: Is there a reasonable way to validate at proposal time whether we'll be able to get the
+		// exported data into a format that the new model can import?  If we can determine it early, then
+		// clients with old ModelLoaders can use that opportunity to dispose early and try to get new
+		// ModelLoaders.
+		assert(this._migratedModel !== undefined, "this._migratedModel should be defined");
+		if (this._migratedModel.supportsDataFormat(this._exportedData)) {
+			// If the migrated model already supports the data format, go ahead with the migration.
+			this._transformedData = this._exportedData;
```

Comment:
@ChumpChief, looks like we will need to remove this check if we want to have the data transform take place before we created the detached model. Unless there is a way to do this check without creating the detached model?

Reply:
Good point, this should stay here - I may have been mixing exportedData with transformedData in my earlier comments.

Reply:
Makes sense. Just to confirm - After the export data step the next steps should be:
1. Load detached model
2. generate transformed data
3. import the transformed data into the detached model

Reply:
That sounds right.  That grouping of 3 things may need to restart fresh if there's an error (since the detached model may be tainted with partial import).

Reply:
Taken care of in the most recent commit!